### PR TITLE
Adding PET-PEESE & WAAP-WLS

### PR DESCRIPTION
**Pull request by FBartos**
adds WAAP-WLS & PET-PEESE analyses
improves documentation in selection models
fixes standard errors of transformed effect size estimates not being transformed to the original scale in cases (in cases when correlations were supplied as input)
Also separates the transformations from the selection models, so they can be shared with the WAAP-WLS & PET-PEESE.

### CODE REVIEW

**Comment by FBartos** [comment]
and I updated themeJasp to themeJaspRaw as raised by Don in another PR

File: R/selectionmodels.R [hunk 1]
```diff
@@ -894,32 +887,32 @@ SelectionModels <- function(jaspResults, dataset, options, state = NULL) {
   plot <- plot + ggplot2::theme(
     axis.ticks.y = ggplot2::element_blank()
   )
-  
-  plot <- jaspGraphs::themeJasp(plot, sides = "b")
+
+  plot <- plot + jaspGraphs::geom_rangeframe(sides = "b") + jaspGraphs::themeJaspRaw()
```
Comment (koenderks) [inline R/selectionmodels.R:891]: Same as above

File: R/transformations.R [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    "general"     = dataset[, options[["inputSE"]]],
+    "correlation" = .maComputeSe(dataset[, options[["inputES"]]], dataset[, options[["inputN"]]], options[["muTransform"]])
+  ))
+}
+.maGetInputPVal            <- function(dataset, options) {
+  # weightfunc uses one-sided p-values as input!
+  if (options[["inputPVal"]] == "")
+    return(pnorm(.maGetInputEs(dataset, options) /.maGetInputSe(dataset, options), lower.tail = FALSE))
+  else
+    return(dataset[, .v(options[["inputPVal"]])])
```
Comment (koenderks) [inline R/transformations.R:25]: there is a `.v()` here

File: R/waapwls.R [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    .wwppFit(jaspResults, dataset, options, "waapWls")
+  }
+
+  # main summary tables
+  .wwppMakeTestsTables(jaspResults, dataset, options, "waapWls")
+  .wwppMakeEstimatesTables(jaspResults, dataset, options, "waapWls")
+
+
+  # figures
+  if (options[["plotModels"]])
```
Comment (koenderks) [inline R/waapwls.R:35]: Not necessary, but you could think about moving `if(options[["plotModels"]]` inside the `wwppEstimatesPlot` function (and others)

File: inst/help/petpeese.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+### Plots
+---
+#### Meta-Regression Estimate
+Visualizes the estimated meta-regression of the relationship between the effect sizes and standard errors.
+
+#### Mean model estimates
+Visualizes mean effect size estimates from all fitted models.
+
+
+### References
```
Comment (koenderks) [inline inst/help/petpeese.md:45]: I think we use a bullet list as reference section, see below (from descriptives): 
image

File: inst/help/selectionmodels.md [hunk 5]
```diff
@@ -1,18 +1,18 @@
-Robust Bayesian Meta-Analysis
+Selection Models
 ===
 
-The selection models allow the user to specify a publication bias adjusted random and fixed effects meta-analytic models. The options allow the user to specify the selection process using either one or two-sided p-value cutoffs, which are internally transformed to one-sided p-value cutoffs prior to the fitting process.
+The selection models are meta-analytic methods on the fixed and random effects meta-analytic models that adjust for publication bias operating on p-values (Iyengar & Greenhouse, 1988; Vevea & Hedges, 1995). The analysis allow users to specify the selection process using either one or two-sided p-value cutoffs, which are internally transformed to one-sided p-value cutoffs prior to the fitting process.
```
Comment (koenderks) [inline inst/help/selectionmodels.md:4]: Maybe remove `The` at the start of the sentence? Maybe I am reading this wrong though

File: tests/testthat/test-petpeese.R [hunk 6]
```diff
@@ -0,0 +1,149 @@
+context("Meta Analysis - PET-PEESE")
```
Comment (koenderks) [inline tests/testthat/test-petpeese.R:1]: Is it possible to add the (commented out) R code and output of these results in the unit tests? Not required though.

File: R/waapwlspetpeese.R [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+  if (options[["inputN"]] != "")
+    .hasErrors(dataset               = dataset,
+               type                  = c("negativeValues"),
+               negativeValues.target = options[["inputN"]],
+               exitAnalysisIfErrors  = TRUE)
+
+
+  if (options[["inputES"]] != "" && options[["measures"]] == "correlation")
+    if (any(dataset[, options[["inputES"]]] <= -1) || any(dataset[, options[["inputES"]]] >= 1))
+      .quitAnalysis(gettextf("Error in %s: All correlation coefficients need to be between -1 and 1.", options[["inputES"]]))
```
Comment (koenderks) [inline R/waapwlspetpeese.R:76]: Maybe apply the same principle to this error message.

File: R/waapwlspetpeese.R [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+    }
+
+    models[["object"]] <- list(
+      pet   = summaryPet,
+      peese = summaryPeese
+    )
+
+  }
+
+  return()
```
Comment (koenderks) [inline R/waapwlspetpeese.R:233]: The `return()` is not necessary right?

File: R/waapwlspetpeese.R [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+
+  models   <- jaspResults[["models"]]$object
+
+  ### test of effect
+  effectTest <- createJaspTable(title = gettext("Test of Effect"))
+  effectTest$position <- 1
+  fitTests[["effectTest"]] <- effectTest
+
+  effectTest$addColumnInfo(name = "type",  title = "",            type = "string")
+  effectTest$addColumnInfo(name = "stat",  title = gettext("t"),  type = "number")
```
Comment (koenderks) [inline R/waapwlspetpeese.R:315]: These things do not need a `gettext()` right? `t` and `p` that is

File: R/waapwlspetpeese.R [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+
+  ### test of bias
+  if (type == "petPeese") {
+    biasTest <- createJaspTable(title = gettext("Test of Publication Bias"))
+    biasTest$position <- 2
+    fitTests[["biasTest"]] <- biasTest
+
+    biasTest$addColumnInfo(name = "type",  title = "",                type = "string")
+    biasTest$addColumnInfo(name = "stat",  title = gettext("t"),      type = "number")
+    biasTest$addColumnInfo(name = "df",    title = gettext("df"),     type = "integer")
```
Comment (koenderks) [inline R/waapwlspetpeese.R:362]: Same here

File: R/waapwlspetpeese.R [hunk 11]
```diff
@@ -0,0 +1,610 @@
...
+    estimates$dependOn(c(.wwppDependencies, "estimates"))
+    jaspResults[["estimates"]] <- estimates
+  } else {
+    estimates <- jaspResults[["estimates"]]
+  }
+
+  # mean estimates
+  if (is.null(estimates[["mean"]]) && options[["estimatesMean"]]) {
+    estimatesMean <- createJaspTable(title = gettextf(
+      "Mean Estimates(%s)",
```
Comment (koenderks) [inline R/waapwlspetpeese.R:397]: This could have a space between Estimates and (%s), but maybe this is on purpose.

File: R/waapwlspetpeese.R [hunk 12]
```diff
@@ -0,0 +1,610 @@
...
+      if (options[["measures"]] == "correlation") "\u03C1" else "\u03BC"
+    ))
+    estimatesMean$position <- 1
+    estimates$dependOn(c("estimatesMean"))
```
Comment (koenderks) [inline R/waapwlspetpeese.R:401]: Does not require a `c()`.

File: R/waapwlspetpeese.R [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+    estimates[["estimatesMean"]] <- estimatesMean
+    estimatesMean <- .wwppFillEstimates(jaspResults, estimatesMean, models, options, type)
+  }
+
+  # PET-PEESE estimates
+  if (type == "petPeese") {
+    if (is.null(estimates[["petPeese"]]) && options[["estimatesPetPeese"]]) {
+      petPeese <- createJaspTable(title = gettext("PET-PEESE Regression Estimates"))
+      petPeese$position  <- 2
+      petPeese$dependOn(c("estimatesPetPeese"))
```
Comment (koenderks) [inline R/waapwlspetpeese.R:411]: same here

File: R/waapwlspetpeese.R [hunk 14]
```diff
@@ -0,0 +1,610 @@
...
+      estimates[["petPeese"]] <- petPeese
+      petPeese <- .wwppFillPetPeese(jaspResults, petPeese, models, options)
+    }
+  }
+
+  # heterogeneity estimates
+  if (is.null(estimates[["heterogeneity"]]) && options[["estimatesSigma"]]) {
+    heterogeneity <- createJaspTable(title = gettext("Multiplicative Heterogeneity Estimates"))
+    heterogeneity$position <- 3
+    heterogeneity$dependOn(c("estimatesSigma"))
```
Comment (koenderks) [inline R/waapwlspetpeese.R:421]: here

File: R/waapwlspetpeese.R [hunk 15]
```diff
@@ -0,0 +1,610 @@
...
+    estimates[["heterogeneity"]] <- heterogeneity
+    heterogeneity <- .wwppFillHeterogeneity(jaspResults, heterogeneity, models, options, type)
+  }
+
+  return()
```
Comment (koenderks) [inline R/waapwlspetpeese.R:426]: No `return()` required.

File: R/waapwlspetpeese.R [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+      options[["muTransform"]],
+      "cohensD"  = gettext("Cohen's d"),
+      "fishersZ" = gettext("Fisher's z"))
+  else
+    yLabel <- gettext("Effect Size")
+
+  # make the plot happen
+  plot <- ggplot2::ggplot()
+  # mean
+  plot <- plot + ggplot2::geom_polygon(
```
Comment (koenderks) [inline R/waapwlspetpeese.R:468]: You can just use `+` here instead of `plot <- plot + `

File: R/waapwlspetpeese.R [hunk 17]
```diff
@@ -0,0 +1,610 @@
...
+    ggplot2::aes(
+      x = fitSeSequence,
+      y = fitPrediction[,"fit"]
+    ),
+    size = 1.25)
+
+  plot <- plot + ggplot2::scale_x_continuous(
+    gettext("Standard Error"),
+    breaks = jaspGraphs::getPrettyAxisBreaks(c(0, fitSe)),
+    limits = range(jaspGraphs::getPrettyAxisBreaks(c(0, fitSe))),
```
Comment (koenderks) [inline R/waapwlspetpeese.R:485]: What is now limits is equal to `fitSeRange` right? Maybe just make the breaks and let limits be `range(xBreaks)` and breaks `xBreaks`.

File: R/waapwlspetpeese.R [hunk 18]
```diff
@@ -0,0 +1,610 @@
...
+    oob    = scales::oob_keep)
+  plot <- plot + ggplot2::scale_y_continuous(
+    yLabel,
+    breaks = jaspGraphs::getPrettyAxisBreaks(c(fitPrediction[,"fit"], fitEs)),
```
Comment (koenderks) [inline R/waapwlspetpeese.R:489]: Not entirely the same, but also better to make the breaks separately

File: R/waapwlspetpeese.R [hunk 19]
```diff
@@ -0,0 +1,610 @@
...
+    limits = range(jaspGraphs::getPrettyAxisBreaks(c(fitPrediction[,"fit"], fitEs))),
+    oob    = scales::oob_keep)
+  plot <- plot + ggplot2::geom_point(
```
Comment (koenderks) [inline R/waapwlspetpeese.R:492]: Any particular reason why you use this instead of `jaspGraphs::geom_point()`?

File: R/waapwlspetpeese.R [hunk 20]
```diff
@@ -0,0 +1,610 @@
...
+      x  = fitSe,
+      y  = fitEs),
+    size  = 2,
+    shape = 18
+  )
+
+  plot <- plot + jaspGraphs::geom_rangeframe() + jaspGraphs::themeJaspRaw()
+  plotRegression$plotObject <- plot
+
+  return()
```
Comment (koenderks) [inline R/waapwlspetpeese.R:503]: No return required here

File: R/waapwlspetpeese.R [hunk 21]
```diff
@@ -0,0 +1,610 @@
...
+  plot   <- plot + ggplot2::geom_point(
+    ggplot2::aes(
+      x = estimates[,"mean"],
+      y = 1:nrow(estimates)),
+    shape = 15)
+  plot <- plot + ggplot2::geom_line(ggplot2::aes(x = c(0,0), y = c(.5, nrow(estimates) + 0.5)), linetype = "dotted")
+  plot <- plot + ggplot2::scale_x_continuous(
+    bquote("Mean Estimate"~.(if (options[["measures"]] == "correlation") bquote(rho) else bquote(mu))),
+    breaks = jaspGraphs::getPrettyAxisBreaks(range(c(0, estimates[,"lowerCI"], estimates[,"upperCI"]))),
+    limits = range(jaspGraphs::getPrettyAxisBreaks(range(c(0, estimates[,"lowerCI"], estimates[,"upperCI"])))))
```
Comment (koenderks) [inline R/waapwlspetpeese.R:564]: Same applies for the breaks here

File: R/waapwlspetpeese.R [hunk 22]
```diff
@@ -0,0 +1,610 @@
...
+  summaryFit$uCI <- summaryFit[, "Estimate"] + qnorm(0.975) * summaryFit[, "Std. Error"]
+
+  return(summaryFit)
+}
+.wwppAddWaapFootnote         <- function(table, waap) {
+
+  if (is.null(waap))
+    return()
+  else if (length(waap) == 0)
+    table$addFootnote(symbol = gettext("Warning:") , gettextf("There were only %1$i adequatelly powered studies. WAAP was not estimated.", attr(waap, "nPowered")))
```
Comment (koenderks) [inline R/waapwlspetpeese.R:607]: Maybe put WAAP was not estimated before "there were only...". Not required.

**Review by koenderks — CHANGES_REQUESTED** [review]
I made some comments about what I think can be improved about the readability of the code. It might be worth to look at some of these points, but not all of them are equally important. If you are ready for another round, let me know!

File: R/petpeese.R [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+PetPeese <- function(jaspResults, dataset, options, state = NULL) {
```
[the remark this answers is not in the document]
Comment (FBartos) [reply R/petpeese.R:18]: It does, apart from the main analysis functions right? At least that's what I saw in the rest of the Meta-Analysis module.

File: R/selectionmodels.R [hunk 24]
```diff
@@ -87,136 +85,134 @@ SelectionModels <- function(jaspResults, dataset, options, state = NULL) {
...
   if (options[["inputES"]] != "" && options[["measures"]] == "correlation")
-    if (any(dataset[, .v(options[["inputES"]])] <= -1) || any(dataset[, .v(options[["inputES"]])] >= 1))
+    if (any(dataset[, options[["inputES"]]] <= -1) || any(dataset[, options[["inputES"]]] >= 1))
       .quitAnalysis(gettextf("Error in %s: All correlation coefficients need to be between -1 and 1.", options[["inputES"]]))
-  
-  
-  
+
+  if (options[["inputN"]] != "" && any(dataset[, options[["inputN"]]] < 4))
+    .quitAnalysis(gettextf("Error in %s: All sample sizes need to be larger than 4.", options[["inputN"]]))
```
[the remark this answers is not in the document]
Comment (FBartos) [reply R/selectionmodels.R:117]: yes, changed, thanks!

File: R/selectionmodels.R [hunk 25]
```diff
@@ -798,25 +791,24 @@ SelectionModels <- function(jaspResults, dataset, options, state = NULL) {
...
     breaks = xSteps,
     labels = xTics,
     limits = c(0, 1))
   plot <- plot + ggplot2::scale_y_continuous(
     gettext("Publication Probability"),
-    breaks = y_tics,
-    limits = range(y_tics))
-  
-  plot <- jaspGraphs::themeJasp(plot)
+    breaks = yTics,
```
[the remark this answers is not in the document]
Comment (FBartos) [reply R/selectionmodels.R:802]: yes, changed

File: R/transformations.R (hunk 2 above)
Comment (FBartos) [reply R/transformations.R:25]: ah, last forgotten one :D thanks, removed

File: R/waapwls.R (hunk 3 above)
Comment (FBartos) [reply R/waapwls.R:35]: keeping it outside since I like it for readibility

File: inst/help/petpeese.md (hunk 4 above)
Comment (FBartos) [reply inst/help/petpeese.md:45]: changed

File: inst/help/selectionmodels.md (hunk 5 above)
Comment (FBartos) [reply inst/help/selectionmodels.md:4]: No, you are reading it correctly. It's just my bad English.

File: R/waapwlspetpeese.R (hunk 7 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:76]: done

File: R/waapwlspetpeese.R (hunk 8 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:233]: no, but I thought that it was preferred to write them explicitly

File: R/waapwlspetpeese.R (hunk 9 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:315]: yes, removed

File: R/waapwlspetpeese.R (hunk 11 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:397]: it was not, good catch, thanks!

File: R/waapwlspetpeese.R (hunk 12 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:401]: fixed there and elsewhere

File: R/waapwlspetpeese.R (hunk 15 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:426]: yes, but I would like to keep it here as well :)

File: R/waapwlspetpeese.R (hunk 16 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:468]: fixed here and in other places (hope I didn't miss any)

File: R/waapwlspetpeese.R (hunk 17 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:485]: true, changed

File: R/waapwlspetpeese.R (hunk 21 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:564]: changed

File: R/waapwlspetpeese.R (hunk 22 above)
Comment (FBartos) [reply R/waapwlspetpeese.R:607]: yes, that's indeed better. Thanks

File: tests/testthat/test-petpeese.R (hunk 6 above)
Comment (FBartos) [reply tests/testthat/test-petpeese.R:1]: ok, I added as in issue to be addressed later for now: https://github.com/jasp-stats/INTERNAL-jasp/issues/1560

File: R/selectionmodels.R [hunk 26]
```diff
@@ -461,25 +456,25 @@ SelectionModels <- function(jaspResults, dataset, options, state = NULL) {
...
   heterogeneityTest$position <- 1
   fitTests[["heterogeneityTest"]] <- heterogeneityTest
-  
-  heterogeneityTest$addColumnInfo(name = "stat",  title = gettext("Q"),  type = "number")
-  heterogeneityTest$addColumnInfo(name = "df",    title = gettext("df"), type = "integer")
-  heterogeneityTest$addColumnInfo(name = "pVal",  title = gettext("p"),  type = "pvalue")
-  
+
+  heterogeneityTest$addColumnInfo(name = "stat",  title = "Q",  type = "number")
+  heterogeneityTest$addColumnInfo(name = "df",    title = "df", type = "integer")
```
Comment (koenderks) [inline R/selectionmodels.R:471]: `df` needs a `gettext()`

File: R/selectionmodels.R [hunk 27]
```diff
@@ -518,29 +513,29 @@ SelectionModels <- function(jaspResults, dataset, options, state = NULL) {
...
   fitTests[["biasTest"]] <- biasTest
-  
-  biasTest$addColumnInfo(name = "type",  title = "",                type = "string")
-  biasTest$addColumnInfo(name = "stat",  title = gettext("ChiSq"),  type = "number")
-  biasTest$addColumnInfo(name = "df",    title = gettext("df"),     type = "integer")
-  biasTest$addColumnInfo(name = "pVal",  title = gettext("p"),      type = "pvalue")
-  
+
+  biasTest$addColumnInfo(name = "type",  title = "",       type = "string")
+  biasTest$addColumnInfo(name = "stat",  title = "ChiSq",  type = "number")
```
Comment (koenderks) [inline R/selectionmodels.R:524]: `ChiSq` might need a gettext(), can this be different in other languages?

File: R/selectionmodels.R [hunk 28]
```diff
@@ -518,29 +513,29 @@ SelectionModels <- function(jaspResults, dataset, options, state = NULL) {
...
+  biasTest$addColumnInfo(name = "df",    title = "df",     type = "integer")
```
Comment (koenderks) [inline R/selectionmodels.R:525]: `df` needs a `gettext()`

**Review by koenderks — CHANGES_REQUESTED** [review]
Just some minor changes. In all files I believe that `df` and `ChiSq` might need `gettext()`'s. Looks fine to me after that!

**Comment by FBartos** [comment]
Ah ok, should be fixed now. Thanks! :)

**Review by koenderks — CHANGES_REQUESTED** [review]
Okay, some last changes (to both analyses) and then this can be merged.

1. Unicode is not shown in WAAP-WLS plot
image

2. I think `Transform correlations` should be capitalized.
image

3. Any particular reason why this is not in alphabetical order?
image

4. Any particular reason why you do not use `jaspGraphs::geom_point()` in these graphs to make them JASP style?
image
Like this maybe:
image

**Comment by FBartos** [comment]
1) I can't reproduce it. Can it be OS specific? I'm on Windows and generating the label in this way:
```
ggplot2::scale_x_continuous(
      bquote("Mean Estimate"~.(if (options[["measures"]] == "correlation") bquote(rho) else bquote(mu))),
      breaks = xTicks,
      limits = range(xTicks))
```
image

2) I don't think so, it's a ``label`` not a ``title``
```
		DropDown
		{
			visible:	measuresCorrelation.checked
			label:		qsTr("Transform correlations")
			name:		"muTransform"
			values:
			[
				{ label: qsTr("Fisher's z"),	value: "fishersZ"},
				{ label: qsTr("Cohen's d"),		value: "cohensD"}
			]
		}
```

3) Not sure that alphabetical is of the highest importance here. A "normal" Meta-Analysis should be on top. It makes sense to have PET-PEESE bellow WAAP-WLS since it's a more complex model. However, I can see Selection models going under PET-PEESE instead of being above WAAP-WLS.

4) Oh, they were jaspGraphs, I just forgot that I set the size and shape manually :D. Fixed now.

**Comment by koenderks** [comment]
About the first point, I know that on Windows unicode works here, but maybe @vandenman can help out here? Could just be my windows machine.

**Comment by vandenman** [comment]
@koenderks could you try a different font? This could just be windows + locale differences. If it works on one windows machine I'm happy though.

Also, I was not aware that the capitalization rules are determined by whether the QML components has property `title` or `label`. It would make a lot of sense to do that, but IIRC then Bruno initially mixed these a lot and Tim would actually prefer it if everything was `title`. This would be a good reason to have differently named properties though.

**Comment by FBartos** [comment]
> Also, I was not aware that the capitalization rules are determined by whether the QML components has property title or label. It would make a lot of sense to do that, but IIRC Bruno initially mixed these a lot and Tim would actually prefer it if everything was title. This would be a good reason to have differently named properties though.

I think that probably Tim told me this in one of the Mixed Model PRs a year ago? I'm not completely sure who&when&why though since it was more than a year ago.

**Comment by EJWagenmakers** [comment]
I agree that ordering by simplicity is better than alphabetically.

**Comment by FBartos** [comment]
I'm moving selection models below PET-PEESE then, so we first have the "normal" meta-analytic stuff, and then the publication bias adjustment stuff in order of increasing complexity

**Review by koenderks — APPROVED** [review]
A different font does not fix this issue. Maybe look at this in the bug fixing phase
image

**Pull request merged**